### PR TITLE
Extract out service logic from StreamDao to a new Service Layer

### LINKED ISSUE #122: Extract out service logic from  StreamDao to a new Service Layer
# Desired Behavior
Extract out service logic from  StreamDao to a new Service Layer

# Benefits
* All service logic remains in Service layer, and UnitTesting is easy 
* All DB calls move to DAO layer

### PR DESCRIPTION
**Pull request by prabhakar1983**
# stream-registry PR

fixed #122 

### Changed
* Renamed the ProducerDao, ConsumerDao, RegionDao, and ClusterDao as ProducerService, ConsumerService, RegionService and CluserService respectively.
* Extracted out the DB calls using ManagedKStreams and ManagerKafkaProducer (get/getAll/upsert Stream) moved to a new Interface StreamDao.

# PR Checklist Forms

- [x] CHANGELOG.md updated
- [x] Reviewer assigned
- [x] PR assigned (presumably to submitter)
- [x] Labels added (enhancement, bug, documentation)

### CODE REVIEW

File: assembly/src/main/jib/app/stream-registry-config.yaml [hunk 1]
```diff
@@ -11,11 +11,7 @@ env: ${STREAM_REGISTRY_ENVIRONMENT:-local}
 
 # list (yaml syntax) of Jersey Filter Full Class Names loaded at runtime.
 requestFilterClassNames:
-#  - com.homeaway.streamplatform.extensions.filters.YourRequestFilter1
-#  - com.homeaway.streamplatform.extensions.filters.YourRequestFilter2
```
Comment (neoword) [inline assembly/src/main/jib/app/stream-registry-config.yaml]: you sure you want to remove these?

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java [hunk 2]
```diff
@@ -77,6 +70,12 @@
 import com.homeaway.streamplatform.streamregistry.resource.ClusterResource;
 import com.homeaway.streamplatform.streamregistry.resource.RegionResource;
 import com.homeaway.streamplatform.streamregistry.resource.StreamResource;
+import com.homeaway.streamplatform.streamregistry.service.ClusterService;
+import com.homeaway.streamplatform.streamregistry.service.RegionService;
+import com.homeaway.streamplatform.streamregistry.service.StreamClientService;
+import com.homeaway.streamplatform.streamregistry.service.StreamService;
+import com.homeaway.streamplatform.streamregistry.service.impl.*;
+import com.homeaway.streamplatform.streamregistry.service.impl.ClusterServiceImpl;
```
Comment (neoword) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java]: we need to standardize on an import order. (#17 is related)

Comment (neoword) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java]: Please remove the .* imports from your IDE that should only be used for statics and only in tests.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    @Override
+    public Stream getStream(String streamName) throws StreamNotFoundException {
+        log.info("Pulling stream information from global state-store for streamName={}", streamName);
+        Optional<AvroStream> streamValue = streamDao.getStream(streamName).getValue();
+
+        if (!streamValue.isPresent()) {
+            throw new StreamNotFoundException(String.format("Stream=%s not found.", streamName));
+        } else {
```
Comment (neoword) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java]: un-necessary else

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+    @Override
+    public void deleteStream(String streamName) throws StreamNotFoundException{
+        Pair<AvroStreamKey, Optional<AvroStream>> keyValue = streamDao.getStream(streamName);
+
+        if (!keyValue.getValue().isPresent()) {
+            throw new StreamNotFoundException(String.format("Stream=%s not found.", streamName));
+        }
+        try {
+            streamDao.upsertStream(keyValue.getKey(), null);
```
Comment (neoword) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java]: Do you really want the service to swallow all exceptions ? Shouldn't runtime exceptions be bubbled up to the resource layer to be handled as 500s?

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    @Override
+    public List<Stream> getAllStreams() {
+        List<Stream> streamList = new ArrayList<>();
+        log.info("Pulling stream information from local instance's state-store");
+        streamDao.getAllStreams().forEachRemaining(avroStream -> streamList.add(AvroToJsonDTO.convertAvroToJson(avroStream.value)));
+        return streamList;
+    }
+
+    @Override
+    public boolean validateSchemaCompatibility(Stream stream) throws SchemaValidationException {
```
Comment (neoword) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java]: really need to fix this signature... either it should return boolean and not throw an exception
OR return void and throw an exception... not both return booland AND throw an exception.

Comment (neoword) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java]: My preference... boolean (and not throw an exception)

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java [hunk 6]
```diff
@@ -296,50 +294,50 @@ private ManagedKafkaProducer createManagedKafkaProducer(final StreamRegistryConf
...
 
         KafkaManager kafkaManager = new KafkaManagerImpl();
         SchemaManager schemaManager = loadSchemaManager(configuration);
         StreamValidator streamValidator = createStreamValidator(configuration, environment);
 
-        streamDao = new StreamDaoImpl(managedKafkaProducer, managedKStreams, env, regionDao, clusterDao, infraManager, streamValidator, schemaManager,
-            kafkaManager);
-        producerDao = new ProducerDaoImpl(managedKafkaProducer, managedKStreams, env, regionDao, clusterDao, infraManager);
-        consumerDao = new ConsumerDaoImpl(managedKafkaProducer, managedKStreams, env, regionDao, clusterDao, infraManager);
+        streamDao = new StreamDaoImpl(managedKafkaProducer, managedKStreams);
```
Comment (prabhakar1983) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java]: Created a new StreamDaoImpl with managedKafkaProducer and managedKStreams to extract out the calls.

```
    void upsertStream(AvroStreamKey key, AvroStream stream);
    Pair<AvroStreamKey, Optional<AvroStream>> getStream(String streamName);
    KeyValueIterator<AvroStreamKey, AvroStream> getAllStreams();
```

**Comment by neoword** [comment]
make sure to rebase / merge master (CHANGELOG got updated)

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java (hunk 2 above)
Comment (prabhakar1983) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java]: fixed it.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+import com.homeaway.digitalplatform.streamregistry.AvroStream;
+import com.homeaway.digitalplatform.streamregistry.AvroStreamKey;
+import com.homeaway.digitalplatform.streamregistry.ClusterValue;
+import com.homeaway.digitalplatform.streamregistry.OperationType;
+import com.homeaway.streamplatform.streamregistry.configuration.KafkaProducerConfig;
+import com.homeaway.streamplatform.streamregistry.db.dao.KafkaManager;
+import com.homeaway.streamplatform.streamregistry.db.dao.StreamDao;
+import com.homeaway.streamplatform.streamregistry.dto.AvroToJsonDTO;
+import com.homeaway.streamplatform.streamregistry.dto.JsonToAvroDTO;
+import com.homeaway.streamplatform.streamregistry.exceptions.*;
```
Comment (neoword) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java]: import .*

File: assembly/src/main/jib/app/stream-registry-config.yaml (hunk 1 above)
Comment (prabhakar1983) [reply assembly/src/main/jib/app/stream-registry-config.yaml]: That was a merge issue. I fixed that.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java (hunk 3 above)
Comment (prabhakar1983) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java]: Handled the case.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java (hunk 4 above)
Comment (prabhakar1983) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java]: When i refactored i missed it. 
Fixed it now. Throwing back all the Runtime Exceptions and Resource class is equipped to handle it properly

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java (hunk 5 above)
Comment (prabhakar1983) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/service/impl/StreamServiceImpl.java]: Fixed it.
gone with the throws Exception approach, and made it as an void statement, as we did lot of ground work for this approach in our previous PR

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,7 +6,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [0.4.9] - SNAPSHOT
 ### Changed
-- Updated swagger example values (#139)
+- Updated swagger example values
```
Comment (neoword) [inline CHANGELOG.md]: No. Please keep the (#139) on the previous line (otherwise you are undoing the merge)

**Comment by neoword** [comment]
LGTM

**Pull request merged**